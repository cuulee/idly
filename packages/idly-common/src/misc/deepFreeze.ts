export function deepFreeze<T extends { [index: string]: any }>(
  obj: T,
  freeze = true
): T {
  if (!freeze) {
    return obj;
  }
  // Retrieve the property names defined on obj
  const propNames = Object.getOwnPropertyNames(obj);

  // Freeze properties before freezing self
  propNames.forEach((name: string) => {
    const prop = obj[name];

    // Freeze prop if it is an object
    if (typeof prop === 'object' && prop !== null) {
      deepFreeze(prop);
    }
  });

  // Freeze self (no-op if already frozen)
  return Object.freeze(obj);
}
