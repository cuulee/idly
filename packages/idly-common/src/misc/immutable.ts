// import { Map as ImMap } from 'immutable';

// export ImMap;
// // export const ImList = Immutable.List;
// // export const ImSet = Immutable.Set;
