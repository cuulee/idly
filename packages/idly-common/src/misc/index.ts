export { cancelablePromise, handleErrors } from './cancelablePromise';
export { deepFreeze } from './deepFreeze';
export { isImmutableList } from './isImmutableList';
export { isImmutableSet } from './isImmutableSet';
export { sortObjectKeys } from './sortObjectKeys';
export { weakCache, weakCache2 } from './weakCache';
export { unescape } from './unescape';
export { tileToQuadkey } from './tileToQuadkey';
