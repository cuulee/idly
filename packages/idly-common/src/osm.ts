export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export type ParentWays = Map<NodeId, Set<WayId>>;
// Table used to map id -> entity
export type EntityTable = Map<EntityId, Entity>;

export type Tags = Map<string, string>;

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export type FeatureProps = { [key: string]: string };

export type NodeGeometry =
  | OsmGeometry.POINT
  | OsmGeometry.VERTEX
  | OsmGeometry.VERTEX_SHARED;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  VERTEX_SHARED = 'vertex_shared', // the one shared among multiple ways
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation'
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation'
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: EntityId[];
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: RelationMember[];
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: number;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

export interface RelationMember {
  readonly id: string;
  readonly type: string;
  readonly role: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
