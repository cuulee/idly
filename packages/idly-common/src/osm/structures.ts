export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export interface Tags {
  readonly [key: string]: string;
}

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export interface FeatureProps {
  [key: string]: string;
}

export type NodeGeometry = OsmGeometry.POINT | OsmGeometry.VERTEX;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation',
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation',
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: ReadonlyArray<EntityId>;
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: ReadonlyArray<RelationMember>;
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: string;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

// id & ref are the same thing
// id exists for backward compatibility with iD
// we are using ref
export interface RelationMember {
  readonly id: string;
  readonly ref: string;
  readonly type?: string;
  readonly role?: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
