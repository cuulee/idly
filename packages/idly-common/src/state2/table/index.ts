export {
  OneToManyTable,
  oneToManyTableAddIndex,
  oneToManyTableUpdateIndex,
  oneToManyTableCreate,
  oneToManyTableFilter,
  oneToManyTableRemove,
  oneToManyTableRemoveIndex,
} from './oneToMany';

export {
  Table,
  tableAdd,
  tableUpdate,
  tableBulkAdd,
  tableBulkGet,
  tableBulkRemove,
  tableCreate,
  tableFilter,
  tableFindValue,
  TableFindValueType,
  tableGet,
  tableHas,
  tableMap,
  tableRemove,
  tableReduce,
  tableFindKey,
  tableForEach,
} from './regular';
