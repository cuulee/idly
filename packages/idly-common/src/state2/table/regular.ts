export type Table<T> = Map<string, T>;

export const tableCreate = <T>(): Table<T> => new Map<string, T>();

export const tableAdd = <T>(t: Table<T>, k: string, e: T) => t.set(k, e);

export const tableBulkAdd = <T>(
  t: Table<T>,
  it: IterableIterator<[string, T]> | Iterable<[string, T]>
) => {
  for (const [k, v] of it) {
    tableAdd<T>(t, k, v);
  }
  return t;
};

export const tableRemove = <T>(t: Table<T>, k: string) => t.delete(k);

export const tableBulkRemove = <T>(
  t: Table<T>,
  it: IterableIterator<string> | Iterable<string>
) => {
  for (const k of it) {
    tableRemove<T>(t, k);
  }
  return t;
};
export const tableBulkGet = <T>(t: Table<T>, k: string[]) =>
  k.map(i => tableGet<T>(t, i));

export const tableGet = <T>(t: Table<T>, k: string) => t.get(k);

export const tableHas = <T>(t: Table<T>, k: string) => t.has(k);

export const tableUpdate = <T>(f: (a?: T) => T, t: Table<T>, k: string) =>
  tableAdd<T>(t, k, f(tableGet<T>(t, k)));

export const tableMap = <T, Z>(f: (v: T, k: string) => Z, t: Table<T>) => {
  const result: Z[] = [];
  t.forEach((v, k) => result.push(f(v, k)));
  return result;
};

export const tableReduce = <T, Z>(
  f: (prev: Z, cur: T, key: string) => Z,
  base: Z,
  t: Table<T>
) => {
  for (const [k, v] of t) {
    base = f(base, v, k);
  }
  return base;
};

export const tableValues = <T>(t: Table<T>) => t.values();

export const tableFilter = <T>(
  f: (r: T, id: string) => boolean,
  t: Table<T>
): Table<T> => {
  const result: Table<T> = new Map();
  t.forEach((e, id) => f(e, id) && result.set(id, e));
  return result;
};

export const tableForEach = <T>(
  f: (r: T, id: string) => any,
  t: Table<T>
): Table<T> => {
  t.forEach(f);
  return t;
};

export type TableFindValueType = <T>(
  f: (r: T, id: string) => boolean,
  t: Table<T>
) => T | undefined;

export const tableFindValue: TableFindValueType = (f, t) => {
  for (const [id, e] of t) {
    if (f(e, id)) {
      return e;
    }
  }
  return;
};

export type TableFindKeyType = <T>(
  f: (r: T, id: string) => boolean,
  t: Table<T>
) => string | undefined;
export const tableFindKey: TableFindKeyType = (f, t) => {
  for (const [id, e] of t) {
    if (f(e, id)) {
      return id;
    }
  }
  return;
};
