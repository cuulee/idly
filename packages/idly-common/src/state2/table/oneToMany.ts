import { Identity as I } from 'monet';
import { mapFilterValuesIntoArray, setAddIterable, setCreate } from '../helper';
import { Table, tableGet, tableRemove, tableUpdate } from '../table';

export type OneToManyTable<T> = Table<Set<T>>;
export const oneToManyTableCreate = <T>(): OneToManyTable<T> => new Map();

export const oneToManyTableRemove = <T>(
  table: OneToManyTable<T>,
  index: string
) => tableRemove(table, index);

export const oneToManyTableRemoveIndex = <T>(
  table: OneToManyTable<T>,
  index: string,
  val: T
) =>
  I(tableGet(table, index))
    .map(s => s && s.delete(val))
    .get();

export const oneToManyTableAddIndex = <T>(
  table: OneToManyTable<T>,
  index: string,
  val: T
) =>
  I((s = setCreate()) => s.add(val))
    .map(f => tableUpdate(f, table, index))
    .get();

export const oneToManyTableUpdateIndex = <T>(
  table: OneToManyTable<T>,
  index: string,
  val: T[]
) =>
  I((s = setCreate()) => setAddIterable(s, val))
    .map(f => tableUpdate(f, table, index))
    .get();

export const oneToManyTableFilter = <T>(
  foo: ((v: Set<T>, k: string) => boolean),
  table: OneToManyTable<T>
) => mapFilterValuesIntoArray(foo, table);
