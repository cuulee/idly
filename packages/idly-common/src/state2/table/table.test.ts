import * as fs from 'fs';
import { deepFreeze } from '../../misc/deepFreeze';

import { Entity } from '../../osm/structures';

import { iterableFlattenToSet } from '../helper';
import { tableBulkAdd, tableBulkRemove, tableCreate, tableGet } from '../table';
import {
  OneToManyTable,
  oneToManyTableCreate,
  oneToManyTableFilter,
  oneToManyTableUpdateIndex,
} from './oneToMany';

const smallDataset = (freeze: boolean = false): Array<[string, Entity]> =>
  fs
    .readFileSync('./src/state2/mosm.jsonl', 'utf-8')
    .split('\n')
    .map(r => JSON.parse(r) as any)
    .map(r => (freeze ? deepFreeze(r) : r))
    .map(toIterable);

const toIterable: (r: Entity) => [string, Entity] = r => [r.id, r];

type QuadkeyTable = OneToManyTable<string>;

const filterByQuadkey = (table: QuadkeyTable, quadkey: string) =>
  iterableFlattenToSet(
    oneToManyTableFilter((_, k) => quadkeyFiltering(quadkey, k), table)
  );

const quadkeyFiltering = (quadkey: string, k: string) =>
  k.startsWith(quadkey) || quadkey.startsWith(k);

describe('entityTableAddBulk', () => {
  const table = tableCreate();
  it('should add', () => {
    const t = smallDataset();
    expect(tableBulkAdd(table, t)).toMatchSnapshot();
  });
});

describe('entityTableGet', () => {
  const table = tableCreate();
  tableBulkAdd(table, smallDataset());
  let toMatch: any = smallDataset().find(([id]) => id === 'n12');
  toMatch = toMatch && toMatch[1];
  it('basic', () => {
    expect(tableGet(table, 'n12')).toEqual(toMatch);
  });
});

describe('entityTableRemoveBulk', () => {
  const table = tableCreate();
  tableBulkAdd(table, smallDataset());

  const newTable = tableCreate();
  tableBulkAdd(newTable, smallDataset());
  newTable.delete('n4133782717');
  newTable.delete('n4133782723');

  it('should remove', () => {
    expect(tableBulkRemove(table, ['n4133782717', 'n4133782723'])).toEqual(
      newTable
    );
  });
  it('should not remove anything on empty array', () => {
    expect(tableBulkRemove(table, [])).toEqual(newTable);
  });
  it('should remove ', () => {
    newTable.delete('w411770875');
    expect(tableBulkRemove(table, ['w411770875'])).toEqual(newTable);
  });
});

describe('oneToManyTable', () => {
  it('should get correct quadkeys for empty quadkeys', () => {
    const table = oneToManyTableCreate<string>();
    const dataset = smallDataset().map(r => r[0]);
    oneToManyTableUpdateIndex(table, '', dataset);

    expect(filterByQuadkey(table, '').size).toEqual(dataset.length);
  });

  it('should add correct quadkey', () => {
    const table = oneToManyTableCreate<string>();

    const dataset = smallDataset();
    const count = 7;
    oneToManyTableUpdateIndex(
      table,
      '1',
      smallDataset()
        .map(r => r[0])
        .slice(0, count)
    );
    oneToManyTableUpdateIndex(
      table,
      '3',
      smallDataset()
        .map(r => r[0])
        .slice(count)
    );
    expect(filterByQuadkey(table, '3').size).toBe(dataset.length - count);
    expect(filterByQuadkey(table, '2').size).toBe(0);
    expect(filterByQuadkey(table, '1').size).toBe(count);
    expect(filterByQuadkey(table, '').size).toBe(dataset.length);
  });

  it('should give correct subset quadkey', () => {
    const table = oneToManyTableCreate<string>();

    const dataset = smallDataset();
    const count = 5;
    oneToManyTableUpdateIndex(
      table,
      '13',
      smallDataset()
        .map(r => r[0])
        .slice(0, count)
    );
    oneToManyTableUpdateIndex(
      table,
      '1',
      smallDataset()
        .map(r => r[0])
        .slice(count)
    );
    expect(filterByQuadkey(table, '1').size).toBe(dataset.length);
    expect(filterByQuadkey(table, '').size).toBe(dataset.length);
  });
  it('should give correct subset quadkey', () => {
    const table = oneToManyTableCreate<string>();
    const dataset = smallDataset();
    const count = 5;
    oneToManyTableUpdateIndex(
      table,
      '21',
      smallDataset()
        .map(r => r[0])
        .slice(0, count)
    );
    oneToManyTableUpdateIndex(
      table,
      '12',
      smallDataset()
        .map(r => r[0])
        .slice(count)
    );
    expect(filterByQuadkey(table, '21').size).toBe(count);
    expect(filterByQuadkey(table, '12').size).toBe(dataset.length - count);
  });
  it('should give not give  other  quadkey', () => {
    const table = oneToManyTableCreate<string>();
    const dataset = smallDataset();
    const count = 5;
    oneToManyTableUpdateIndex(
      table,
      '13',
      smallDataset()
        .map(r => r[0])
        .slice(0, count)
    );
    oneToManyTableUpdateIndex(
      table,
      '23',
      smallDataset()
        .map(r => r[0])
        .slice(count)
    );

    expect(filterByQuadkey(table, '3').size).toBe(0);
    expect(filterByQuadkey(table, '2').size).toBe(dataset.length - count);
    expect(filterByQuadkey(table, '1').size).toBe(count);
  });

  it('multiple quadkeys', () => {
    const table = oneToManyTableCreate<string>();

    oneToManyTableUpdateIndex(
      table,
      '1',
      smallDataset()
        .map(r => r[0])
        .slice(0, 4)
    );
    oneToManyTableUpdateIndex(
      table,
      '12',
      smallDataset()
        .map(r => r[0])
        .slice(4, 8)
    );
    oneToManyTableUpdateIndex(
      table,
      '123',
      smallDataset()
        .map(r => r[0])
        .slice(8, 12)
    );
    oneToManyTableUpdateIndex(
      table,
      '223',
      smallDataset()
        .map(r => r[0])
        .slice(12, 14)
    );
    oneToManyTableUpdateIndex(
      table,
      '122',
      smallDataset()
        .map(r => r[0])
        .slice(14)
    );

    expect(filterByQuadkey(table, '').size).toBe(smallDataset().length);
    expect(filterByQuadkey(table, '1').size).toBe(18);
    expect(filterByQuadkey(table, '12').size).toBe(18);
    expect(filterByQuadkey(table, '123').size).toBe(12);
    expect(filterByQuadkey(table, '122').size).toBe(14);
    expect(filterByQuadkey(table, '223').size).toBe(2);
  });
});
