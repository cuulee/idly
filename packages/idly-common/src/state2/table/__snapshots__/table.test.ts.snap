// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entityTableAddBulk should add 1`] = `
Map {
  "n12" => Object {
    "attributes": Object {
      "changeset": "138701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n12",
    "loc": Object {
      "lat": 34.5280866,
      "lon": 74.1795357,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782712" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782712",
    "loc": Object {
      "lat": 34.5280866,
      "lon": 74.1795357,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782713" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782713",
    "loc": Object {
      "lat": 34.528206,
      "lon": 74.1807924,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782714" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782714",
    "loc": Object {
      "lat": 34.5283653,
      "lon": 74.1822907,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782715" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782715",
    "loc": Object {
      "lat": 34.528425,
      "lon": 74.1815174,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782716" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782716",
    "loc": Object {
      "lat": 34.5285445,
      "lon": 74.1788832,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782717" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782717",
    "loc": Object {
      "lat": 34.5290024,
      "lon": 74.1788349,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782718" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782718",
    "loc": Object {
      "lat": 34.5292015,
      "lon": 74.179584,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782719" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:39Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782719",
    "loc": Object {
      "lat": 34.529301,
      "lon": 74.1824357,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782720" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782720",
    "loc": Object {
      "lat": 34.5294603,
      "lon": 74.1794149,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782721" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782721",
    "loc": Object {
      "lat": 34.5296992,
      "lon": 74.1823873,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782722" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782722",
    "loc": Object {
      "lat": 34.5297191,
      "lon": 74.1788349,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782723" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782723",
    "loc": Object {
      "lat": 34.5300377,
      "lon": 74.1821698,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782724" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782724",
    "loc": Object {
      "lat": 34.5303562,
      "lon": 74.1783274,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782725" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782725",
    "loc": Object {
      "lat": 34.530396,
      "lon": 74.1816382,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782726" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782726",
    "loc": Object {
      "lat": 34.5305354,
      "lon": 74.1805024,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782727" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782727",
    "loc": Object {
      "lat": 34.5309137,
      "lon": 74.1799224,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782728" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782728",
    "loc": Object {
      "lat": 34.5311526,
      "lon": 74.1786899,
    },
    "tags": Object {},
    "type": "node",
  },
  "n4133782729" => Object {
    "attributes": Object {
      "changeset": "38701677",
      "timestamp": "2016-04-19T17:58:40Z",
      "uid": "592097",
      "user": "Dombo",
      "version": "1",
      "visible": true,
    },
    "id": "n4133782729",
    "loc": Object {
      "lat": 34.5314114,
      "lon": 74.1793182,
    },
    "tags": Object {},
    "type": "node",
  },
  "w411770875" => Object {
    "attributes": Object {
      "changeset": "50089927",
      "timestamp": "2017-07-06T17:19:04Z",
      "uid": "2815653",
      "user": "Srihari Thalla",
      "version": "2",
      "visible": true,
    },
    "id": "w411770875",
    "nodes": Array [
      "n4133782719",
      "n4133782714",
      "n4133782715",
      "n4133782713",
      "n4133782712",
      "n4133782716",
      "n4133782717",
      "n4133782718",
      "n4133782720",
      "n4133782722",
      "n4133782724",
      "n4133782728",
      "n4133782729",
      "n4133782727",
      "n4133782726",
      "n4133782725",
      "n4133782723",
      "n4133782721",
      "n4133782719",
    ],
    "tags": Object {
      "landuse": "residential",
      "place": "village",
      "source": "MSF Survey",
    },
    "type": "way",
  },
}
`;
